lib/dotcom/trip_plan/itinerary_groups.ex

```diff
@@ -0,0 +1,145 @@
+defmodule Dotcom.TripPlan.ItineraryGroups do
+  @moduledoc """
+  Group itineraries by unique legs.
+
+  A unique leg is defined as a leg that has a unique combination of mode, from, and to.
+  But, this does not include walking legs that are less than 0.2 miles.
+  """
+
+  alias Dotcom.TripPlan.{Itinerary, Leg, PersonalDetail, TransitDetail}
+  alias OpenTripPlannerClient.ItineraryTag
+
+  @spec from_itineraries([Itinerary.t()]) :: [
+          %{itineraries: [Itinerary.t()], summary: map()}
+        ]
+  def from_itineraries(itineraries) do
+    itineraries
+    |> Enum.group_by(&unique_legs_to_hash/1)
+    |> Enum.map(&group_departures/1)
+    |> Enum.reject(&Enum.empty?(&1))
+    |> Enum.sort_by(fn
+      %{itineraries: [%{tag: tag} | _] = _} ->
+        Enum.find_index(ItineraryTag.tag_priority_order(), &(&1 == tag))
+
+      _ ->
+        -1
+    end)
+  end
+
+  defp unique_legs_to_hash(%Itinerary{legs: legs}) do
+    legs
+    |> Enum.reject(&short_walking_leg?/1)
+    |> Enum.map(&unique_leg_to_tuple/1)
+    |> :erlang.phash2()
+  end
+
+  defp unique_leg_to_tuple(%Leg{mode: %PersonalDetail{}} = leg) do
+    {:WALK, leg.from.name, leg.to.name}
+  end
+
+  defp unique_leg_to_tuple(%Leg{mode: %{route: route}} = leg) do
+    {Routes.Route.type_atom(route.type), leg.from.name, leg.to.name}
+  end
+
+  defp group_departures({_hash, grouped_itineraries}) do
+    summarized_legs =
+      grouped_itineraries
+      |> Enum.map(& &1.legs)
+      |> Enum.zip_with(&Function.identity/1)
+      |> Enum.map(fn legs ->
+        legs
+        |> Enum.uniq_by(&combined_leg_to_tuple/1)
+        |> Enum.map(&to_summarized_leg/1)
+        |> Enum.reduce(%{walk_minutes: 0, cents: 0, routes: []}, &summarize_legs/2)
+      end)
+      |> remove_short_intermediate_walks()
+
+    summary =
+      grouped_itineraries
+      |> Enum.map(fn itinerary ->
+        itinerary
+        |> Map.take([:start, :stop, :tag, :duration, :accessible?, :walk_distance])
+        |> Map.put(
+          :total_cost,
+          DotcomWeb.TripPlanView.get_one_way_total_by_type(itinerary, :highest_one_way_fare)
+        )
+      end)
+      |> summarize_itineraries()
+      |> Map.put(:summarized_legs, summarized_legs)
+
+    %{itineraries: ItineraryTag.sort_tagged(grouped_itineraries), summary: summary}
+  end
+
+  defp remove_short_intermediate_walks(summarized_legs) do
+    summarized_legs
+    |> Enum.with_index()
+    |> Enum.reject(fn {leg, index} ->
+      index > 0 && index < Kernel.length(summarized_legs) - 1 &&
+        (leg.routes == [] && leg.walk_minutes < 5)
```

Comment:
Nit (non-blocking):
```suggestion
        (leg.routes == [] && leg.walk_minutes < @short_walk_threshold)
```

So that 5 minutes isn't a magic value hidden away in the middle, but displayed prominently at the top.

I could even see a justification for making it an env variable, but I think that's overkill.